File: frontend/src/types/audit.ts
/**
 * Audit log types for the frontend.
 */

export type ActionType =
  | "create"
  | "update"
  | "complete"
  | "uncomplete"
  | "delete"
  | "recurring_auto_create";

export interface AuditLogEntry {
  id: string;
  entityType: string;
  entityId: string;
  userId: string;
  actionType: ActionType;
  fieldChanged: string | null;
  oldValue: string | null;
  newValue: string | null;
  timestamp: string;
  isSystemAction: boolean;
}

export interface AuditLogListResponse {
  items: AuditLogEntry[];
  total: number;
}

/**
 * Format an action type for display.
 */
export function formatActionType(action: ActionType): string {
  const actionMap: Record<ActionType, string> = {
    create: "Created",
    update: "Updated",
    complete: "Completed",
    uncomplete: "Reopened",
    delete: "Deleted",
    recurring_auto_create: "Auto-created (Recurring)",
  };
  return actionMap[action] || action;
}

/**
 * Get a color class for an action type.
 */
export function getActionColor(action: ActionType): string {
  const colorMap: Record<ActionType, string> = {
    create: "text-green-600 dark:text-green-400",
    update: "text-blue-600 dark:text-blue-400",
    complete: "text-primary",
    uncomplete: "text-amber-600 dark:text-amber-400",
    delete: "text-destructive",
    recurring_auto_create: "text-purple-600 dark:text-purple-400",
  };
  return colorMap[action] || "text-muted-foreground";
}

File: backend/app/models/audit.py
"""Audit log model for tracking all mutations."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import uuid4

import sqlalchemy as sa
from sqlmodel import Column, Field, SQLModel

ActionTypeEnum = sa.Enum(
    "create",
    "update",
    "complete",
    "uncomplete",
    "delete",
    "recurring_auto_create",
    name="actiontype",
    create_type=False,
)



class ActionType(str, Enum):
    """Types of auditable actions."""

    CREATE = "create"
    UPDATE = "update"
    COMPLETE = "complete"
    UNCOMPLETE = "uncomplete"
    DELETE = "delete"
    RECURRING_AUTO_CREATE = "recurring_auto_create"


class AuditLog(SQLModel, table=True):
    """Immutable audit log for tracking all task mutations.

    This table is append-only - no updates or deletes are allowed.
    """

    __tablename__ = "audit_logs"

    # Primary Key
    id: str = Field(
        default_factory=lambda: str(uuid4()),
        primary_key=True,
        max_length=36,
        description="Unique identifier (UUID)",
    )

    # Entity Reference
    entity_type: str = Field(
        max_length=50,
        index=True,
        description="Type of entity (task, tag)",
    )
    entity_id: str = Field(
        max_length=36,
        index=True,
        description="ID of the affected entity",
    )

    # Actor
    user_id: str = Field(
        max_length=36,
        index=True,
        description="User who performed the action",
    )

    # Action Details
    action_type: ActionType = Field(
        sa_column=Column(ActionTypeEnum, nullable=False),
        description="Type of action performed",
    )
    field_changed: Optional[str] = Field(
        default=None,
        max_length=50,
        description="Field that was changed (for updates)",
    )
    old_value: Optional[str] = Field(
        default=None,
        max_length=2000,
        description="Previous value (JSON string)",
    )
    new_value: Optional[str] = Field(
        default=None,
        max_length=2000,
        description="New value (JSON string)",
    )

    # Metadata
    timestamp: datetime = Field(
        default_factory=datetime.utcnow,
        nullable=False,
        index=True,
        description="When the action occurred",
    )
    is_system_action: bool = Field(
        default=False,
        description="True if action was system-generated (e.g., recurring task)",
    )
